R/asis-nodes.R

```diff
@@ -0,0 +1,199 @@
+    xml2::xml_set_attr(bmath[endless], "latex-pair", tags)
+    xml2::xml_set_attr(bmath[headless], "latex-pair", tags)
+    for (i in tags) {
+      fix_partial_inline(i, body, ns)
+    }
+  }
+  
+  copy_xml(body)
+}
+
+fix_partial_inline <- function(tag, body, ns) {
+  # find everything between the tagged pair
+  math_lines <- find_between_inlines(body, ns, tag)
+  # make sure everything between the tagged pair is labeled as 'asis'
+  filling <- math_lines[is.na(xml2::xml_attr(math_lines, "latex-pair"))]
+  set_asis(filling)
+  filling <- xml2::xml_find_all(filling, ".//node()")
+  set_asis(filling)
+  # paste the lines together and create new nodes
+  n <- length(math_lines)
+  char <- as.character(math_lines)
+  char[[1]] <- sub("[$]", "$</text><text asis='true'>", char[[1]])
+  char[[n]] <- sub("[<]text ", "<text asis='true' ", char[[n]])
+  nodes <- paste(char, collapse = "")
+  nodes <- xml2::xml_children(set_default_space(nodes))
+  # add the new nodes to the bottom of the existing math lines 
+  last_line <- math_lines[n]
+  to_remove <- math_lines[-n]
+  add_node_siblings(last_line, nodes, remove = TRUE)
+  # remove the duplicate lines
+  xml2::xml_remove(to_remove)
+}
+
+fix_fully_inline <- function(math) {
+  char <- as.character(math)
+  # Find inline math that is complete and wrap it in text with asis
+  # <text>this is $\LaTeX$ text</text>
+  #   becomes
+  # <text>this is </text><text asis='true'>$\LaTeX$</text><text> text</text>
+  char <- gsub(
+    pattern = inline_dollars_regex("full"),
+    replacement = "</text><text asis='true'>\\1</text><text>", 
+    x = char,
+    perl = TRUE
+  )
+  set_default_space(char)
+}
+
+set_default_space <- function(char) {
+  new_nodes <- char_to_nodelist(char)
+  n <- xml2::xml_find_all(new_nodes, ".//node()")
+  # set space to default to avoid weird formatting (this may change)
+  xml2::xml_set_attr(n, "xml:space", "default")
+  new_nodes
+}
+
+char_to_nodelist <- function(txt) {
+  doc <- glue::glue(commonmark::markdown_xml("{paste(txt, collapse = '\n')}")) 
+  doc <- xml2::read_xml(doc)
+  nodes <- xml2::xml_children(xml2::xml_children(doc))
+  nodes[xml2::xml_name(nodes) != "softbreak"]
+}
+
+
+# BLOCK MATH ------------------------------------------------------------------
+
+find_block_math <- function(body, ns) {
+  find_between(body, ns, pattern = "md:text[contains(text(), '$$')]", include = FALSE)
+}
+
+find_between_inlines <- function(body, ns, tag) {
+  to_find <- "md:text[@latex-pair='{tag}']"
+  find_between(body, ns, pattern = glue::glue(to_find), include = TRUE)
+}
+
+protect_block_math <- function(body, ns) {
+  bm <- find_block_math(body, ns)
+  # get all of the internal nodes
+  bm <- xml2::xml_find_all(bm, ".//descendant-or-self::md:*", ns = ns)
+  set_asis(bm) 
+}
+
+# TICK BOXES -------------------------------------------------------------------
+
+tick_check <- function(body, ns) {
+  predicate <- "starts-with(text(), '[ ]') or starts-with(text(), '[x]')"
```

Comment:
out of curiosity, why doesn't the regex contain the "-"/"*"?

Reply:
Because commonmark has already parsed "-"/"*" as a list item, so we see a hierarchy of list/item/paragraph/text on the xml side.